File: backend/services/llm.py
"""
LLM service using Groq (llama-3.3-70b-versatile).
Handles match analysis and chat with context.
"""

import os
import json
import re
from groq import Groq

_client = None


def _get_client():
    global _client
    if _client is None:
        _client = Groq(api_key=os.getenv("GROQ_API_KEY"))
    return _client


def _generate(prompt: str) -> str:
    """Generate a response from Groq."""
    response = _get_client().chat.completions.create(
        model="llama-3.3-70b-versatile",
        messages=[{"role": "user", "content": prompt}],
        temperature=0.3,
        max_tokens=2048,
    )
    return response.choices[0].message.content.strip()


async def analyze_match(resume_text: str, jd_text: str) -> dict:
    """
    Analyze how well a resume matches a job description.
    Returns structured JSON with score, strengths, gaps, insights.
    """
    prompt = f"""You are an expert recruiter. Analyze how well this resume matches the job description.

RESUME:
{resume_text}

JOB DESCRIPTION:
{jd_text}

Return a JSON object with exactly this structure:
{{
  "score": <number 0-100>,
  "strengths": ["strength 1", "strength 2", ...],
  "gaps": ["gap 1", "gap 2", ...],
  "insights": ["insight 1", "insight 2", ...],
  "summary": "2-3 sentence overall assessment"
}}

Be specific — reference actual skills, years of experience, and qualifications from the resume.
Score should reflect realistic match percentage.
Return ONLY valid JSON, no markdown or extra text."""

    text = _generate(prompt)

    # clean markdown code fences if present
    text = re.sub(r"^```json?\s*", "", text)
    text = re.sub(r"\s*```$", "", text)

    try:
        result = json.loads(text)
    except json.JSONDecodeError:
        result = {
            "score": 50,
            "strengths": ["Could not parse detailed analysis"],
            "gaps": ["Analysis format error"],
            "insights": ["Please try again"],
            "summary": text[:200],
        }

    return result


async def chat_with_context(
    question: str,
    retrieved_chunks: list[str],
    conversation_history: list[dict],
    full_resume: str,
) -> str:
    """
    Answer a question using retrieved resume chunks as context.
    This is the 'Generation' part of RAG.
    """
    context = "\n---\n".join(retrieved_chunks)

    history_text = ""
    if conversation_history:
        recent = conversation_history[-6:]  # last 3 Q&A pairs
        for msg in recent:
            role = "User" if msg["role"] == "user" else "Assistant"
            history_text += f"{role}: {msg['content']}\n"

    prompt = f"""You are a recruiter assistant analyzing a candidate's resume.
Answer the question using ONLY the provided resume sections below.
If the information is not in the provided context, say so honestly.

RETRIEVED RESUME SECTIONS:
{context}

PREVIOUS CONVERSATION:
{history_text if history_text else "None"}

QUESTION: {question}

Give a clear, specific answer citing facts from the resume. Be concise."""

    return _generate(prompt)

File: backend/src/services/llm.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { MatchResult, ChatMessage } from '../types';
import dotenv from 'dotenv';
dotenv.config();

const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY || '');
const model = genAI.getGenerativeModel({ model: 'gemini-2.0-flash' });

export async function analyzeMatch(
  resumeText: string,
  jdText: string
): Promise<MatchResult> {
  const prompt = `You are a recruitment AI. Analyze the resume against the job description and provide a JSON response.

RESUME:
${resumeText}

JOB DESCRIPTION:
${jdText}

Respond ONLY with valid JSON in this exact format:
{
  "score": <number 0-100>,
  "strengths": ["strength 1", "strength 2", ...],
  "gaps": ["gap 1", "gap 2", ...],
  "insights": ["insight 1", "insight 2", ...],
  "summary": "brief overall assessment"
}

Be specific. Reference actual skills, years of experience, and qualifications from the resume.`;

  const result = await model.generateContent(prompt);
  const text = result.response.text();

  // extract JSON from response
  const jsonMatch = text.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error('Failed to parse LLM response as JSON');

  return JSON.parse(jsonMatch[0]) as MatchResult;
}

export async function chatWithContext(
  question: string,
  relevantChunks: string[],
  conversationHistory: ChatMessage[],
  resumeText: string
): Promise<string> {
  const contextStr = relevantChunks
    .map((chunk, i) => `[Retrieved Section ${i + 1}]: ${chunk}`)
    .join('\n\n');

  const historyStr = conversationHistory
    .slice(-6)
    .map(m => `${m.role === 'user' ? 'User' : 'AI'}: ${m.content}`)
    .join('\n');

  const prompt = `You are an AI recruiter assistant. Answer the question using ONLY the retrieved resume sections below. Be specific and cite details from the resume.

RETRIEVED RESUME SECTIONS (from vector search):
${contextStr}

${historyStr ? `CONVERSATION HISTORY:\n${historyStr}\n` : ''}
USER QUESTION: ${question}

Rules:
- Base your answer strictly on the retrieved sections
- If the information isn't in the sections, say so
- Be concise and specific
- Reference exact details (names, dates, skills, institutions)`;

  const result = await model.generateContent(prompt);
  return result.response.text();
}
